File: examples_py/tests/framework/test_delay_buffer.py
#!/usr/bin/env python3

# Copyright 2025 Universidad Politécnica de Madrid
# SPDX-License-Identifier: BSD-3-Clause

"""Unit tests for :class:`examples_py.framework.DelayBuffer`."""

import unittest

from examples_py.framework.delay_buffer import DelayBuffer


class DelayBufferTest(unittest.TestCase):

    def test_empty_returns_none(self) -> None:
        buf: DelayBuffer = DelayBuffer()
        self.assertIsNone(buf.latest_available(0.0))
        self.assertEqual(len(buf), 0)

    def test_delivers_payload_once_its_available_at_elapses(self) -> None:
        buf: DelayBuffer = DelayBuffer()
        buf.push(42, 1.0)

        self.assertIsNone(buf.latest_available(0.5))
        self.assertEqual(len(buf), 1)

        v = buf.latest_available(1.0)
        self.assertEqual(v, 42)
        self.assertEqual(len(buf), 0)
        self.assertIsNone(buf.latest_available(1.0))

    def test_returns_latest_available_and_discards_older_entries(self) -> None:
        buf: DelayBuffer = DelayBuffer()
        buf.push(1, 0.1)
        buf.push(2, 0.2)
        buf.push(3, 0.3)

        v = buf.latest_available(0.25)
        self.assertEqual(v, 2)
        self.assertEqual(len(buf), 1)

        self.assertIsNone(buf.latest_available(0.29))
        v3 = buf.latest_available(0.30)
        self.assertEqual(v3, 3)

    def test_out_of_order_push_is_ignored(self) -> None:
        buf: DelayBuffer = DelayBuffer()
        buf.push(1, 1.0)
        buf.push(2, 0.5)  # out of order

        self.assertEqual(len(buf), 1)
        v = buf.latest_available(1.0)
        self.assertEqual(v, 1)

    def test_clear_drops_pending_entries(self) -> None:
        buf: DelayBuffer = DelayBuffer()
        buf.push(1, 1.0)
        buf.push(2, 2.0)
        self.assertEqual(len(buf), 2)
        buf.clear()
        self.assertEqual(len(buf), 0)
        self.assertIsNone(buf.latest_available(5.0))


if __name__ == '__main__':  # pragma: no cover
    unittest.main()

File: examples_cpp/tests/framework/test_delay_buffer.cpp
// Copyright 2025 Universidad Politécnica de Madrid
// SPDX-License-Identifier: BSD-3-Clause

#include <gtest/gtest.h>

#include "framework/delay_buffer.hpp"

using mpc_examples::framework::DelayBuffer;

TEST(DelayBufferTest, EmptyReturnsNullopt) {
  DelayBuffer<int> buf;
  EXPECT_FALSE(buf.latestAvailable(0.0).has_value());
  EXPECT_EQ(buf.size(), 0u);
}

TEST(DelayBufferTest, DeliversPayloadOnceItsAvailableAtElapses) {
  DelayBuffer<int> buf;
  buf.push(42, 1.0);

  // Before the available_at, the payload is invisible.
  EXPECT_FALSE(buf.latestAvailable(0.5).has_value());
  EXPECT_EQ(buf.size(), 1u);

  // At available_at it becomes visible (and is consumed).
  auto v = buf.latestAvailable(1.0);
  ASSERT_TRUE(v.has_value());
  EXPECT_EQ(v.value(), 42);
  EXPECT_EQ(buf.size(), 0u);

  // Subsequent queries at the same time return nothing (buffer drained).
  EXPECT_FALSE(buf.latestAvailable(1.0).has_value());
}

TEST(DelayBufferTest, ReturnsLatestAvailableAndDiscardsOlderEntries) {
  DelayBuffer<int> buf;
  buf.push(1, 0.1);
  buf.push(2, 0.2);
  buf.push(3, 0.3);

  // At t = 0.25 the buffer should hand back payload "2" (the latest with
  // available_at <= t) and drop "1" + "2".
  auto v = buf.latestAvailable(0.25);
  ASSERT_TRUE(v.has_value());
  EXPECT_EQ(v.value(), 2);
  EXPECT_EQ(buf.size(), 1u);

  // The remaining entry (3) becomes visible later.
  EXPECT_FALSE(buf.latestAvailable(0.29).has_value());
  auto v3 = buf.latestAvailable(0.30);
  ASSERT_TRUE(v3.has_value());
  EXPECT_EQ(v3.value(), 3);
}

TEST(DelayBufferTest, OutOfOrderPushIsIgnored) {
  DelayBuffer<int> buf;
  buf.push(1, 1.0);
  buf.push(2, 0.5);  // out of order: must be a no-op

  EXPECT_EQ(buf.size(), 1u);
  auto v = buf.latestAvailable(1.0);
  ASSERT_TRUE(v.has_value());
  EXPECT_EQ(v.value(), 1);
}

TEST(DelayBufferTest, ClearDropsPendingEntries) {
  DelayBuffer<int> buf;
  buf.push(1, 1.0);
  buf.push(2, 2.0);
  EXPECT_EQ(buf.size(), 2u);

  buf.clear();
  EXPECT_EQ(buf.size(), 0u);
  EXPECT_FALSE(buf.latestAvailable(5.0).has_value());
}
